Draft Implementation of a Practice Mode

mk56-spn opened the pull request:
https://github.com/ppy/osu/discussions/20778, https://github.com/ppy/osu/discussions/20323, https://github.com/ppy/osu/discussions/13076,https://github.com/ppy/osu/discussions/16723, #14629 https://github.com/ppy/osu/discussions/13079, https://github.com/ppy/osu/discussions/13013

Been working on this for almost a month and whilst i still have a lot to do before even id consider this merge-able ( in addition to any requirements added by the core Devs) i feel like this is as far as i can go without bloating the PR, at least until i have some feedback on the existing commits, that said i will list existing known issues here, as well as features that need implementing.

Known issues: 
- [x] practice overlay being unduly dimmed
- [ ] Preview does not play song audio
- [ ] Freeze preview screen gameplay
- [x] Fix beatmapsets with different length beatmaps not working properly,
- [x] Fix transitions between pause overlay and practice overlay, as well as making sure the main game play remains
 paused throughout ( you can no longer lock yourself without access to any menu, other issues still remain)
- [x] Shows auto mod as being used (Due to it using it for the preview), needs hiding 
- [X] Hide judgements on the gameplay preview
- [x] Fix Blocking fail being neccesary due to seek
- [x] Add proper tests
- [x] Practice rangeSlider is not showing appropriate value format
- [ ] autoplay isnt being triggered. ( should it even be? )
- [x] Adress fail overlay softlocking the player
- [x] Adress converts not showing the ruleset that will be played, instead of the original one


<details>
  <summary>Features:</summary>

- [x] Reimplement toggle for practice mode on song select (coded on an old branch, need to fix any issues before adding it into this one, for now practice mode is set to true always)
- [x] add an in game display indicating the range being played
- [x] make the player auto restart upon reaching the customEnd set by user
- [x] add a disclaimer on the practiceplayerloader screen to distinguish it
- [x] Add keybinds

</details>

For anyone interested in doing code review  for this, id recommend going by commit, since i did my best to separate feature additions into their own commits, wherever applicable and where the size of the implementation merited it

*As an aside this aims to implement BASIC FUNCTIONALITY as a starter, so requesting things like the ability to toggle mods is out of the scope of this PR ( i think )*

Accompanying video for anyone interested : 

https://user-images.githubusercontent.com/74463310/203850666-fc295af7-1a65-40fb-9212-f49f20fe4a47.mp4

And last but not least credit to : https://github.com/harshavb and his rewind mod that allowed me to understand some concepts i'd later use for this

mk56-spn committed:
Implement PracticePlayer class and Overlay

mk56-spn committed:
Implement the gameplay preview for the practice overlay

mk56-spn committed:
Add seek slider for gameplay preview

mk56-spn committed:
Add practice buttons to fail and pause overlay, move seek logic out of practicerangeslider

mk56-spn committed:
Improve Tests

mk56-spn committed:
Store custom restart data in player loader

mk56-spn committed:
Add AutoPlay to the gameplayPreview, Hide Cursor where applicable

mk56-spn committed:
Fix Overlay TestScene, Simplify player loading process and add to SongSelect, fix cursor Hiding for gameplay preview

mk56-spn committed:
Fix naming issues, address song select test failures, double white line removed

bdach commented on osu.Game/Overlays/Practice/PracticeOverlayComponents/PracticeGameplayPreview.cs:
```diff
@@ -0,0 +1,69 @@
+            drawableRuleset.Cursor?.FadeTo(0);
+        }
+
+        private Drawable createGameplayComponents() => new ScalingContainer(ScalingMode.Gameplay)
+        {
+            Child = drawableRuleset
+        };
```
Why is this in a scaling container even? It means it will be affected by "gameplay" scaling in settings.

mk56-spn replied:
oversight. will address

bdach commented:
In addition to the above review, I don't really want to be that guy, but calling this an RFC is way premature if I can make this sort of stuff happen in 30 seconds of testing:

https://user-images.githubusercontent.com/20418176/203868792-c868fec9-8b0f-4a6a-bf44-2cc48814e6f2.mp4

95% of what I'm doing in the above video is trying to use the escape key to unforeseen consequences.

mk56-spn commented on osu.Game/Screens/Select/PlaySongSelect.cs:
```diff
@@ -33,6 +34,8 @@ public class PlaySongSelect : SongSelect
 
         private PlayBeatmapDetailArea playBeatmapDetailArea = null!;
 
+        private bool practiceMode => true;
```
> So practice mode is just permanently on now? Why isn't this TODOed at least? This cannot remain in the final version.

I'm aware. Which is why I have it listed in my feature to do list above to reimplement the toggle I had from another branch and add it to song select. This was merely for ease of testing . Should added a todo though you are right

mk56-spn commented:
> In addition to the above review, I don't really want to be that guy, but calling this an RFC is way premature if I can make this sort of stuff happen in 30 seconds of testing:
> 
> https://user-images.githubusercontent.com/20418176/203868792-c868fec9-8b0f-4a6a-bf44-2cc48814e6f2.mp4
> 
> 95% of what I'm doing in the above video is trying to use the escape key to unforeseen consequences.

Yeah perhaps RFC wasn't the right term here. And I'm aware of this. I have it listed above as "fixing transitions" . Sorry for the troubles

mk56-spn commented on osu.Game/Overlays/Practice/PracticeOverlay.cs:
```diff
@@ -0,0 +1,103 @@
+
+            double? lastTime = drawableRuleset?.Objects.Last().StartTime;
+
+            playerLoader.CustomStart.ValueChanged += startPercent =>
+            {
+                preview.SeekTo(startPercent.NewValue * lastTime!.Value);
+            };
+            playerLoader.CustomEnd.ValueChanged += endPercent =>
+            {
+                preview.SeekTo(endPercent.NewValue * lastTime!.Value);
+            };
```
Wasn't aware of this. Thanks

mk56-spn committed:
Remove need for PracticePlayer.cs DI in overlay and gameplay preview.

Remove dependency on base players "playableBeatmap"

Create clones of PracticePlayerLoader binds in PracticeOverlay.cs

Other small fixes.

mk56-spn committed:
Add SongSelect toggle for practice mode.

peppy commented on osu.Game/Overlays/Practice/PracticeOverlayComponents/PracticeRangeSliderComponent.cs:
```diff
@@ -0,0 +1,170 @@
+
+namespace osu.Game.Overlays.Practice.PracticeOverlayComponents
+{
+    public class PracticeRangeSliderComponent : CompositeDrawable
```
thisclass seems like mostluy copy paste of another class. please don't do this.

mk56-spn replied:
yep i was planning to merge them together ( had trouble doing it initially)  ill add it to the todo list :)

peppy commented on osu.Game/Overlays/Practice/PracticePlayer.cs:
```diff
@@ -0,0 +1,45 @@
+        {
+            var playableBeatmap = beatmap.Value.GetPlayableBeatmap(beatmap.Value.BeatmapInfo.Ruleset);
+
+            SetGameplayStartTime(loader.CustomStart.Value * (playableBeatmap.HitObjects.Last().StartTime - playableBeatmap.HitObjects.First().StartTime));
```
why is this done using DI? the loader creates the player so you can call directly and avoid the copmlexity.

mk56-spn replied:
sorry for the late response, was attempting to fix it somehow but im out of ideas, to put it simply i think this is the only case where the the playerloader ( well, practiceplayerloader here ) needs to be able to pass data to the player being created which doesnt exist until after the playerloader itself has been created ( except for ofcourse the base playerloader class which has access to more stuff) so i opted to DI the playerloader into the player after creation to feed it the data, instead of altering the base playerloader to allow its derived classes to create the player from within themselves.

peppy commented on osu.Game.Tests/Visual/Gameplay/TestScenePracticeOverlay.cs:
```diff
@@ -0,0 +1,26 @@
+    public class TestScenePracticeOverlay : TestSceneAllRulesetPlayers
+    {
+        [Cached]
+        private PracticePlayerLoader loader { get; set; } = new PracticePlayerLoader();
```
how is this test even usable to test anything? i don't understand

mk56-spn replied:
i need to work on it, originally there was no link for the overlay through song select so i was using this test for everything, i gutted it later on due to it causing me issues, i plan to make the full fledged test once i have the way to call the overlay better conceptualised

peppy commented:
Just looking at the list of issues makes me think this should not be opened yet. Unless you're looking for help with someone. If that's the case then please add commentary on what you need help with.

mk56-spn commented:
Opened it as a draft mainly because i was having trouble gauging the dev teams desired direction for many parts of the implementation, opening this was more so a way of making sure the general direction I'm taking is appropriate

I also do have a few issues id appreciate help with but seeing as I've caught several mentions of excessive burden due to too much to review i can wait / keep grinding at it myself. 

Incase you/ someone else find a spare moment to give me a helping hand here a thing i cannot seem to wrap my head around:

why is the overlay ( practiceoverlay ) seemingly  dimming itself?, ive tried searching through parent classes and its probably obvious but i don't understand whats governing this behavior.

and again sorry if the opening of the PR seems to hasty, i'm confident i can pull through with the PR without much external help, since IMO the hardest parts are done?

bdach commented:
> why is the overlay ( practiceoverlay ) seemingly dimming itself?, ive tried searching through parent classes and its probably obvious but i don't understand whats governing this behavior.

You're not calling [`RegisterBlockingOverlay()`](https://github.com/ppy/osu/blob/a841d35ebd4ecab69ef268273038c04c0432b895/osu.Game/OsuGame.cs#L212-L227). Compare [song select and the mod select overlay](https://github.com/ppy/osu/blob/a841d35ebd4ecab69ef268273038c04c0432b895/osu.Game/Screens/Select/SongSelect.cs#L304).

mk56-spn committed:
Make practice overlay disposable ( fixes dimming issue )

mk56-spn committed:
Merge branch 'master' into Practice_overlay_clean

mk56-spn committed:
Address the new requirement for the partial keyword in DI related classes

mk56-spn committed:
Merge branch 'master' into Practice_overlay_clean

mk56-spn committed:
Remove practice range slider in favour of new generic range slider implementation, fix few remaining partial class issues, fix some overlay issues, adress the need for blocking fail

mk56-spn committed:
alter the playerloader to allow practicePlayerLoader to access its createPlayer function directly, restructure the practicePlayer

mk56-spn committed:
improve seek functionality, improve fail blocking with grace period, isolate seek time logic to preview

mk56-spn committed:
Fix seek not working on beatmapsets with different map lengths, make customEnd trigger restart

mk56-spn committed:
Add display for range being played, disable customEnd temporarily

mk56-spn committed:
Create proper test suite, other minor fixes

mk56-spn commented:
I'm probably going to need some help with the remaining two listed known issues above (, i'm struggling to find a way to freeze the gameplaypreview ( practicegameplaypreview.cs , line 43 ,for the clock currently being used) 

namely: 

am i using the right clock? 

if so how can i set its rated without getting an exception;




also appreciate knowing if i should be testing more scenarios than those i currently am (c1428fg for the pertinent tests)

mk56-spn committed:
Add practice mode indicator to practice player loader

Feodor0090 commented:
This has the same issue as editor test - you are thrown to the center of the map without time to prepare and with all previous objects missed due to fast-forward. IMO both things should trim objects in skipped parts and start player as usually (but with skipped "intro") to let gameplay start as it should, without touching clocks.

Also, shouldn't this be done as mod rather than a switch in footer?

mk56-spn commented:
> This has the same issue as editor test - you are thrown to the center of the map without time to prepare and with all previous objects missed due to fast-forward. IMO both things should trim objects in skipped parts and start player as usually (but with skipped "intro") to let gameplay start as it should, without touching clocks.
> 
> Also, shouldn't this be done as mod rather than a switch in footer?

it was specifically request that it be a switch and not a mod so no, the issue with the clocks is totally unrelated to what you are talking about, and involves freezing the preview, as for being thrown in abruptly, i agree, but im not sure what the path ahead is yet

mk56-spn committed:
Hide HUD whilst in PracticeOverlay, tweak some minor visual aspects

Feodor0090 commented:
- "Practice!" button must be triggerable by enter key or something like that.
- Would be nice if range slider would handle arrow keys somehow without hover. Not sure how to do this, given that we have 2 nubs instead of one. Maybe via tabbing between start and end.
- Fail. Press "practice" in menu. Close the overlay. This will bring player to that state:
![изображение](https://user-images.githubusercontent.com/53872073/205315727-91e489af-b396-4cd0-ab9c-cc88f684497c.png)
Pressing "continue" will softlock player screen.
- Scores are saved, but their dates are in bad state.
![изображение](https://user-images.githubusercontent.com/53872073/205317216-32b64b29-fef1-4213-a410-d1bc194f7dab.png)
- Using keyboard 0-length practice range can be created. Seems to break nothing, but it shouldn't be allowed, should it?

> as for being thrown in abruptly, i agree, but im not sure what the path ahead is yet

Beatmap is reloaded once you choose a range. Can you just filter out hitobjects before start time via `.Where()` at load moment?

mk56-spn commented:
> * "Practice!" button must be triggerable by enter key or som



> * "Practice!" button must be triggerable by enter key or something like that.
> 
>     * Would be nice if range slider would handle arrow keys somehow without hover. Not sure how to do this, given that we have 2 nubs instead of one. Maybe via tabbing between start and end.
> 
>     * Fail. Press "practice" in menu. Close the overlay. This will bring player to that state:
>       ![изображение](https://user-images.githubusercontent.com/53872073/205315727-91e489af-b396-4cd0-ab9c-cc88f684497c.png)
>       Pressing "continue" will softlock player screen.
> 
>     * Scores are saved, but their dates are in bad state.
>       ![изображение](https://user-images.githubusercontent.com/53872073/205317216-32b64b29-fef1-4213-a410-d1bc194f7dab.png)
> 
>     * Using keyboard 0-length practice range can be created. Seems to break nothing, but it shouldn't be allowed, should it?
> 
> 
> > as for being thrown in abruptly, i agree, but im not sure what the path ahead is yet
> 
> Beatmap is reloaded once you choose a range. Can you just filter out hitobjects before start time via `.Where()` at load moment?

Yeah i know keybindings are missing for the practice button. will address now, was busy with footer and other stuff

arrow keys for the slider seem impractical do to for now, i dont intend the range slider as it exists right now to be the permanent seeking tool, but a more advanced version would bloat this PR and it works for now 

ill add a condition to the fail screen, practice mod needs some form of endscreen after an attempt to show you how youve done i feel, but again i dont want this to become 1.3k lines, as i said its a basic implementation , i dont think this stage will be shown to the general public, it can be merged with the song select button removed so it can be all done by parts, ill set it to restart on exist if you try and leave the overlay whilst failed for now 

as for filtering out the objects i don't think its possible the way the base player class drawableruleset creation flow is set up right now, implementation detail would probably be out of the scope of this PR , as its not specific to this as you said, the editor player has the same problem. ill look into it a bit more though

mk56-spn committed:
Stop pause overlay being called if player has failed

mk56-spn committed:
Allow setting format in range slider (for Nubs)

Feodor0090 commented:
- Converts preview are shown in original ruleset, while gameplay use proper one.
- Close + continue:

https://user-images.githubusercontent.com/53872073/206746407-49fcb38a-1670-467c-9734-528e97f4669a.mp4

- Autoplay + close + continue works weird too. It blocks failing but still allows you to play.

https://user-images.githubusercontent.com/53872073/206746729-c5921d58-268d-481f-a136-72ef4d7337e3.mp4

The score?
![изображение](https://user-images.githubusercontent.com/53872073/206746882-91778cd0-85ae-46f6-8ebf-432c2db41084.png)

mk56-spn commented:
> * Converts preview are shown in original ruleset, while gameplay use proper one.
> 
>     * Close + continue:
> 
> 
> 2022-12-09.19-13-25.mp4
> 
>     * Autoplay + close + continue works weird too. It blocks failing but still allows you to play.
> 
> 
> 2022-12-09.19-17-43.mp4
> 
> The score? ![изображение](https://user-images.githubusercontent.com/53872073/206746882-91778cd0-85ae-46f6-8ebf-432c2db41084.png)
still working on score, same for autoplay, i took a break from this for a week or so. eitherway thanks for the additional feedback, some things i wont change for now ( score ) since i want to know if scores should submit locally or not be created at all, until i get confirmation in either direction from the devs ill deal with other things

aware of the converts issue, more complicated than youd think since some custom rulesets dont have autoplay, whether showing the non converted map or nothing at all is preferable is something ill have to think about

mk56-spn committed:
Add support for toggling the practice button in PracticeOverlay.cs with keybindings (enter)

Hacky solution to making customEnd work

Feodor0090 commented:
> some custom rulesets dont have autoplay

I'm quite sure that i played all known customs at least once and the only ruleset which i remember lacking proper autoplay is karaoke. If i'm wrong, what about non-converted maps for such rulesets? This is still an issue even when ignoring converts. It should be workarounded somehow or just ignored for now.

Also, quick way to lose your skin layout:

https://user-images.githubusercontent.com/53872073/206753919-6213dd32-271a-489b-be65-6f1166ad2a84.mp4

mk56-spn committed:
Make preview use the converted ruleset

mk56-spn committed:
Merge branch 'master' into Practice_overlay_clean

mk56-spn committed:
Update practice range counter design

peppy committed:
Merge branch 'master' into Practice_overlay_clean

mk56-spn commented:
fwiw i wouldnt knock myself out reviewing this right now, i have this kind of on hold since song select feels more important, and i plan to do a clean up pass after that, up to you of course, just letting you know

mk56-spn committed:
Add PracticePercentageCounter.cs to HUDOverlay.cs top right elements

Fix some naming.

Remove unnecessary actions

mk56-spn committed:
Add some localisations

mk56-spn committed:
adjust syntax for PracticePercentageCounter.cs and schedule hiding in a more appropriate way

mk56-spn commented:
this PR is a mess by now but its served well in detecting most of the issues the implementation  had,  will open a fresh PR with commits neatly laid out  and perhaps split the gameplay preview into its own PR since its likely going to be used elsewhere as well

mk56-spn committed:
force SkinEditorOverlay.cs to close when  PracticeOverlay.cs becomes visible

peppy commented:
@mk56-spn a few usability concerns while I was briefly checking this branch:

- Gameplay should pause while seeking (and potentially always paused by default) when selecting a start / end point. Adding a play/pause button would be fine if you think there's users that would want to play back in this mode.
- When already practicing a section, the pause menu shows "Practice" when it should probably show "Change practice range"
- The range selection should probably show time in mm:ss in addition to %, to allow more precision in selection.

Probably more, but these are the pieces I recall.

mk56-spn commented:
> @mk56-spn a few usability concerns while I was briefly checking this branch:
> 
>     * Gameplay should pause while seeking (and potentially always paused by default) when selecting a start / end point. Adding a play/pause button would be fine if you think there's users that would want to play back in this mode.
> 
>     * When already practicing a section, the pause menu shows "Practice" when it should probably show "Change practice range"
> 
>     * The range selection should probably show time in mm:ss in addition to %, to allow more precision in selection.
> 
> 
> Probably more, but these are the pieces I recall.

oh yep have those documented ( minus the menu setting rename ) theres plenty more to do though.

 like i only have a vague idea of  what im going to do with rulesets that are incapable of autoplay, although theres no official ones that do this so i think simply throwing  a soft error message ( like the skin editor does for screens that dont support skinning ) should suffice for now? , ( fwiw i tested osu ! tau and that works fine).
 
 gameplay preview needs robust testing... 
 
 code quality is not up to scratch even by my standards now...
 
 in regards to the starting at a more precise ms, i was perhaps considering note snapped start points? 
 
 im not sure about what to do with the lag spikes the seek causes on marathons ( editor has this too ).. 
 
 but yeah plenty for me to go through
